storage/redis/peer_store.go

```diff
@@ -0,0 +1,787 @@
+				numWant--
+			}
+		}
+	}
+
+	APResult := ""
+	for _, pr := range peers {
+		APResult = fmt.Sprintf("%s Peer:[ID: %s, IP: %s(AddressFamily: %s), Port %d]", APResult, pr.ID.String(), pr.IP.String(), IPver, pr.Port)
+	}
+	log.Debug("storage: AnnouncePeers result", log.Fields{
+		"peers": APResult,
+	})
+
+	return
+}
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	resp.InfoHash = ih
+	IPver := addressFamily.String()
+	encoded_ih := ih.String()
+	encoded_ih_L := IPver + "_L_" + encoded_ih // key
+	encoded_ih_S := IPver + "_S_" + encoded_ih // key
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechers_len, err := conn.Do("HLEN", encoded_ih_L)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encoded_ih_L,
+			"error": err,
+		})
+		return
+	}
+	l_len := leechers_len.(int64)
+
+	seeders_len, err := conn.Do("HLEN", encoded_ih_S)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encoded_ih_S,
+			"error": err,
+		})
+		return
+	}
+	s_len := seeders_len.(int64)
+
+	if l_len == 0 && s_len == 0 {
+		return
+	}
+
+	resp.Incomplete = uint32(l_len)
+	resp.Complete = uint32(s_len)
+
+	return
+}
+
+// collectGarbage deletes all Peers from the PeerStore which are older than the
+// cutoff time.
+//
+// This function must be able to execute while other methods on this interface
+// are being executed in parallel.
+func (ps *peerStore) collectGarbage(cutoff time.Time) error {
+	select {
+	case <-ps.closed:
+		return nil
+	default:
+	}
+
+	shards := [2]string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	cutoffUnix := cutoff.UnixNano()
+	start := time.Now()
+
+	for _, shard := range shards {
+		infohashes_list, err := conn.Do("HKEYS", shard) // key
+		if err != nil {
+			return err
+		}
+		infohashes := infohashes_list.([]interface{})
+
+		for _, ih := range infohashes {
+			ih_str := string(ih.([]byte))
+
+			ih_list, err := conn.Do("HGETALL", ih_str) // field
+			if err != nil {
+				return err
+			}
+			con_ih := ih_list.([]interface{})
+
+			if len(con_ih) == 0 {
+				_, err := conn.Do("DEL", ih_str)
+				if err != nil {
+					return err
+				}
+				log.Debug("storage: Deleting Redis", log.Fields{"Hkey": ih_str})
+				_, err = conn.Do("HDEL", shard, ih_str)
+				if err != nil {
+					return err
+				}
+				log.Debug("storage: Deleting Redis", log.Fields{
+					"Hkey":   shard,
+					"Hfield": ih_str,
+				})
+				continue
+			}
+
+			var pk serializedPeer
+			for index, ih_field := range con_ih {
+				if index%2 != 0 { // value
+					mtime, err := strconv.ParseInt(string(ih_field.([]byte)), 10, 64)
+					if err != nil {
+						return err
+					}
+					if mtime <= cutoffUnix {
```

Comment:
Same question as above - what if someone changed the mtime between the `HGETALL` and this check?

Reply:
It doesn't affect the overall Peers conditions as above.